planner/core/find_best_task.go

```diff
@@ -1962,30 +1962,37 @@ func (ds *DataSource) getOriginalPhysicalIndexScan(prop *property.PhysicalProper
 }
 
 func (p *LogicalCTE) findBestTask(prop *property.PhysicalProperty, planCounter *PlanCounterTp) (t task, cntPlan int64, err error) {
-	if !prop.IsEmpty() {
+	if !prop.IsEmpty() && !prop.CanAddEnforcer {
 		return invalidTask, 1, nil
 	}
-	if p.cte.cteTask != nil {
+	var pcte *PhysicalCTE
+	if p.cte.physicalCTE != nil {
 		// Already built it.
-		return p.cte.cteTask, 1, nil
-	}
-	sp, _, err := DoOptimize(context.TODO(), p.ctx, p.cte.optFlag, p.cte.seedPartLogicalPlan)
-	if err != nil {
-		return nil, 1, err
-	}
-
-	var rp PhysicalPlan
-	if p.cte.recursivePartLogicalPlan != nil {
-		rp, _, err = DoOptimize(context.TODO(), p.ctx, p.cte.optFlag, p.cte.recursivePartLogicalPlan)
+		pcte = p.cte.physicalCTE
+	} else {
+		sp, _, err := DoOptimize(context.TODO(), p.ctx, p.cte.optFlag, p.cte.seedPartLogicalPlan)
 		if err != nil {
 			return nil, 1, err
 		}
-	}
 
-	pcte := PhysicalCTE{SeedPlan: sp, RecurPlan: rp, CTE: p.cte, cteAsName: p.cteAsName}.Init(p.ctx, p.stats)
-	pcte.SetSchema(p.schema)
-	t = &rootTask{pcte, sp.statsInfo().RowCount, false}
-	p.cte.cteTask = t
+		var rp PhysicalPlan
+		if p.cte.recursivePartLogicalPlan != nil {
+			rp, _, err = DoOptimize(context.TODO(), p.ctx, p.cte.optFlag, p.cte.recursivePartLogicalPlan)
+			if err != nil {
+				return nil, 1, err
+			}
+		}
+
+		pcte = PhysicalCTE{SeedPlan: sp, RecurPlan: rp, CTE: p.cte, cteAsName: p.cteAsName}.Init(p.ctx, p.stats)
+		pcte.SetSchema(p.schema)
+		p.cte.physicalCTE = pcte
+	}
+	t = &rootTask{pcte, pcte.SeedPlan.statsInfo().RowCount, false}
+	if !prop.IsEmpty() {
+		// When got here, `prop.CanAddEnforcer` must be true,
+		// so we just enforce property.
+		t = enforceProperty(prop, t, pcte.basePlan.ctx)
```

Comment:
We may fail to find the optimal plan for `with recursive cte1(c1) as (select c1 from t1) select /*+ MERGE_JOIN(dt1, dt2) */ * from cte1 dt1 join t2 dt2 on dt1.c1 = dt2.c1` if `t1` has an index on `c1`, i.e, for the cte part, using index on `c1` may be better than enforcing a `Sort` here, but we are unaware of this sort property when calling `DoOptimize` for the cte definitions.

Reply:
For now, the optimization of `seed part` is seperated. So I record this problem in another issue(#25674)